[wasm] Avoid unnecessary relinking when publishing a blazor project for AOT

radical opened the pull request:
`Publish` target triggers `Build` target to run. When publishing, we don't want
to run relinking step during the `Build`, as it will be run for `Publish`
anyway. Earlier there wasn't a good way to differentiate the two cases of
`build` target - `dotnet build`, and `dotnet publish`, but now the sdk sets
`$(_IsPublishing)=true`, which can be used here.

- VS uses `$(DeployOnBuild)`, and that is still supported

- And add tests to run blazor+AOT

radical committed:
[wasm] Wasm.Build.Tests: Add blazor+aot run test

radical committed:
cleanup

radical committed:
[wasm] Wasm.Build.Tests: Use $(_IsPublishing) to avoid unncessary

.. relinking during "Build", when publishing.

`Publish` target triggers `Build` target to run. When publishing, we
don't want to run relinking step during the `Build`, as it will be run
for `Publish` anyway. Earlier there wasn't a good way to differentiate
the two cases of `build` when building, and `build` when publishing, but
now the sdk sets `$(_IsPublishing)=true`, which can be used here.

radical committed:
fix blazor tests

radical committed:
Wasm.Build.Tests: Add support for running published blazor projects

`dotnet run` uses the regular build output. Published blazor app, like
when using AOT, needs to be run more "manually" by starting a web server
in the publish folder. This is accomplished here by using the new
xharness command `wasm webserver`.

radical committed:
Detect runtime assertions in blazor runs

radical commented:
Depends on https://github.com/dotnet/xharness/pull/1001

radical commented:
This adds missing tests that run AOT'ed blazor app. And that repros the currently broken state.

radical committed:
Update test to track changes

radical committed:
cleanup

maraf reviewed (approved):
Looks good to me 👍

radical committed:
Disable BlazorPublishRunTest because of https://github.com/dotnet/runtime/issues/82481

radical committed:
Update test to track changes

pavelsavara commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -172,7 +173,7 @@
     <!-- Use a unique property, so the already run wasm targets can also run -->
     <MSBuild Projects="$(MSBuildProjectFile)"
```
I'm curious how this works.
Do we run separate MSBuild process here (which doesn't share variables and items from parent ?)
Do we still need to have it as separate process ?
Do we still need `_WasmInNestedPublish_UniqueProperty_XYZ` ?

radical replied:
> Do we run separate MSBuild process here (which doesn't share variables and items from parent ?)

No, but it runs it in a fresh context. I will be removing this, but I'm waiting for the blazor<->runtime build changes to happen.

> Do we still need to have it as separate process ?

It's not necessarily a separate process. And this PR doesn't change anything about this.

> Do we still need _WasmInNestedPublish_UniqueProperty_XYZ ?

Yep

Essentially, this PR doesn't change any behavior w.r.t the nested publish. The nested publish bit was a hacky solution, and will be removed. But I'm holding that off till we get further on the build consolidation with blazor.

pavelsavara replied:
@ilonatommy opened issue for ICU related dependency on not nesting it 
https://github.com/dotnet/runtime/issues/82819

radical committed:
Update xharness dependency

Updating 'Microsoft.DotNet.XHarness.CLI': '1.0.0-prerelease.23117.1' => '1.0.0-prerelease.23151.1' (from build '20230301.1' of 'https://github.com/dotnet/xharness')
Updating 'Microsoft.DotNet.XHarness.TestRunners.Common': '1.0.0-prerelease.23117.1' => '1.0.0-prerelease.23151.1' (from build '20230301.1' of 'https://github.com/dotnet/xharness')
Updating 'Microsoft.DotNet.XHarness.TestRunners.Xunit': '1.0.0-prerelease.23117.1' => '1.0.0-prerelease.23151.1' (from build '20230301.1' of 'https://github.com/dotnet/xharness')

radical commented:
Added the xharness dependency update here, from https://github.com/dotnet/runtime/pull/82600

radical commented:
The failures are unrelated and https://github.com/dotnet/runtime/issues/82868 .
